### PR TITLE
7 send historical device data via websocket

### LINKED ISSUE #7: Send historical device data via websocket
See https://github.com/bifravst/Muninn-backend/issues/47

### PR DESCRIPTION
**Pull request by pudkrong**

### CODE REVIEW

File: features/HistoricalRequest.feature.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+And I store the converted device messages of device ID `${deviceId}` into
+timestream using `${ts}` and decrement the timestamp by `0.5` minutes as this
+JSON
+
+```json
+[
+  {
+    "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/gain",
+    "mA": 3.40141
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]: add timestamp property here, as the device would send it using JSONata expressions.

File: features/HistoricalRequest.feature.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+And I store
+`$toMillis($fromMillis($millis(), '[Y9999]-[M99]-[D99]T[H99]:[m]:00Z'))` into
+`ts` and generate a series with decrement of `1` minute each time for `5`
+iterations
+
+And I store the converted device messages of device ID `${deviceId}` into
+timestream using `${ts}` and decrement the timestamp by `0.5` minutes as this
+JSON
+
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]: Remove this.

File: features/HistoricalRequest.feature.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    "ts": ${ts},
+    "message": "gain",
+    "attributes": {
+      "avgMA": { "attribute": "mA", "aggregate": "avg" }
+    }
+  }
+}
+```
+
+Then wait for `5` seconds
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]: Remove, use `Soon` keyword.

File: features/HistoricalRequest.feature.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+Then the historical response should equal to this JSON
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]:
```suggestion
Soon the historical response should equal to this JSON
```

File: features/HistoricalRequest.feature.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        "mA": 3.28449,
+        "ts": ${ts4}
+      },
+      {
+        "mA": 3.24925,
+        "ts": ${ts5}
+      }
+    ]
+  },
+  "ts": ${ts}
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]:
```suggestion

```

Should be server time.

**Review by coderbyheart — CHANGES_REQUESTED** [review]

File: features/HistoricalRequest.feature.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+When I send websocket request as this JSON
+
+```json
+{
+  "message": "message",
+  "payload": {
+    "@id": "${requestId}",
+    "type": "lastHour",
+    "ts": ${ts},
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]:
```suggestion

```

Remove.

File: features/HistoricalRequest.feature.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+And I store `$join(['request-', $formatBase($random() * 10000000, 16)])` into
+`requestId`
+
+When I send websocket request as this JSON
+
+```json
+{
+  "message": "message",
+  "payload": {
+    "@id": "${requestId}",
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]: Message must have `@context`, see https://github.com/bifravst/Muninn-backend/issues/47#issuecomment-1592567975

**Review by coderbyheart — CHANGES_REQUESTED** [review]

File: features/HistoricalRequest.feature.md [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+### Examples
+
+| appId | messageType | data    | deductMsFromTS | storeName |
+| ----- | ----------- | ------- | -------------- | --------- |
+| SOLAR | DATA        | 3.40141 | 0              | ts1       |
+| SOLAR | DATA        | 3.75718 | 30000          | ts2       |
+| SOLAR | DATA        | 3.73368 | 60000          | ts3       |
+| SOLAR | DATA        | 3.58041 | 90000          | ts4       |
+| SOLAR | DATA        | 3.24925 | 120000         | ts5       |
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]: Should follow established format of devices publishing messages:

```suggestion
Given I store `ts - ${deductMsFromTS}` into `pastTs`

And the device `${fingerprint:deviceId}` publishes this message to the topic
`m/d/${fingerprint:deviceId}/d2c`

```json
{
  "appId": "${appId}",
  "messageType": "DATA",
  "ts": ${pastTs},
  "data": "${data}"
}
```

### Examples

| appId | messageType | data    | deductMsFromTS |
| ----- | ----------- | ------- | -------------- |
| SOLAR | DATA        | 3.40141 | 0              |
| SOLAR | DATA        | 3.75718 | 30000          |
| SOLAR | DATA        | 3.73368 | 60000          |
| SOLAR | DATA        | 3.58041 | 90000          |
| SOLAR | DATA        | 3.24925 | 120000         |

```

File: features/HistoricalRequest.feature.md [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+```json
+{
+  "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/historical-data",
+  "@id": "420eac59-5ce8-4751-b7d1-217811382095",
+  "attributes": {
+    "avgMA": [
+      {
+        "mA": 3.40141,
+        "ts": ${ts1}
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]: No need to use store, these ts can be calculated from the know start `ts` and the expected bin.

File: historicalData/historicalDataRepository.spec.ts [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	Context: jest.fn().mockReturnValue('model'),
+}))
+
+describe('transformTimestreamData', () => {
+	it('should extract single attribute from multiple values in single row', () => {
+		const data = [
+			{
+				avgMA: 1,
+				maxMA: 9,
+				minMA: 0,
```
Comment (coderbyheart) [inline historicalData/historicalDataRepository.spec.ts]: Always use real values, so there are no suprises. mA is float.

File: feature-runner/lib/websocket.ts [hunk 11]
```diff
@@ -41,6 +42,15 @@ export const createWebsocketClient = ({
...
 			messages,
+			send: async (message) => {
+				return new Promise((resolve, reject) => {
+					client.send(JSON.stringify(message), (error) => {
+						if (error) return reject(error)
+
+						resolve(void 0)
+					})
+				})
+			},
```
Comment (coderbyheart) [inline feature-runner/lib/websocket.ts]:
```suggestion
			send: async (message) => new Promise<void>((resolve, reject) => {
					client.send(JSON.stringify(message), (error) => {
						if (error) return reject(error)
						resolve()
					})
				}),
```

File: lambda/onMessage.ts [hunk 12]
```diff
@@ -1,26 +1,51 @@
...
 
 export const handler = async (
 	event: AuthorizedEvent,
 ): Promise<APIGatewayProxyStructuredResultV2> => {
 	log.info('event', { event })
-
 	await repo.extendTTL(event.requestContext.connectionId)
 
+	if (event.body !== undefined) {
+		const { payload } = JSON.parse(event.body)
```
Comment (coderbyheart) [inline lambda/onMessage.ts:44]: Validate payload

File: feature-runner/steps/device.ts [hunk 13]
```diff
@@ -31,16 +34,18 @@ const createDeviceForModel =
...
-		const match =
-			/^I have the fingerprint for a `(?<model>[^`]+)` device in `(?<storageName>[^`]+)`$/.exec(
-				step.title,
-			)
+		const match = matchGroups(
+			Type.Object({
+				model: Type.String(),
+				storageName: Type.String(),
+			}),
+		)(
```
Comment (coderbyheart) [inline feature-runner/steps/device.ts]: Oh nice, I forgot about that!

File: feature-runner/steps/device.ts [hunk 14]
```diff
@@ -180,11 +189,83 @@ const publishDeviceMessage =
 		})
 	}
 
+const publishDeviceMessages =
```
Comment (coderbyheart) [inline feature-runner/steps/device.ts]: Remove.

File: feature-runner/steps/device.ts [hunk 15]
```diff
@@ -31,16 +34,18 @@ const createDeviceForModel =
 	}: StepRunnerArgs<
 		World & Record<string, string>
 	>): Promise<StepRunResult> => {
-		const match =
```
Comment (coderbyheart) [inline feature-runner/steps/device.ts]: Unrelated change.

File: feature-runner/steps/device.ts [hunk 16]
```diff
@@ -31,16 +34,18 @@ const createDeviceForModel =
...
+			Type.Object({
+				model: Type.String(),
+				storageName: Type.String(),
+			}),
+		)(
+			/^I have the fingerprint for a `(?<model>[^`]+)` device in `(?<storageName>[^`]+)`$/,
+			step.title,
+		)
 		if (match === null) return noMatch
-		const { model, storageName } = match.groups as {
```
Comment (coderbyheart) [inline feature-runner/steps/device.ts]: Unrelated change.

File: feature-runner/steps/device.ts [hunk 17]
```diff
@@ -100,17 +105,20 @@ const getDevice =
 		},
 		context: { devicesTable },
 	}: StepRunnerArgs<World>): Promise<StepRunResult> => {
-		const match = /^The device id `(?<key>[^`]+)` should equal$/.exec(
```
Comment (coderbyheart) [inline feature-runner/steps/device.ts]: Unrelated change.

File: feature-runner/steps/device.ts [hunk 18]
```diff
@@ -100,17 +105,20 @@ const getDevice =
...
-		)
+		const match = matchGroups(
+			Type.Object({
+				key: Type.String(),
+			}),
+		)(/^The device id `(?<key>[^`]+)` should equal$/, step.title)
+
 		if (match === null) return noMatch
 
-		progress(`Get data with id ${match.groups?.key} from ${devicesTable}`)
```
Comment (coderbyheart) [inline feature-runner/steps/device.ts]: Unrelated change.

File: feature-runner/steps/device.ts [hunk 19]
```diff
@@ -100,17 +105,20 @@ const getDevice =
...
+		progress(`Get data with id ${match.key} from ${devicesTable}`)
 		const res = await db.send(
 			new GetItemCommand({
 				TableName: devicesTable,
 				Key: {
-					deviceId: { S: match.groups?.key ?? '' },
```
Comment (coderbyheart) [inline feature-runner/steps/device.ts]: Unrelated change.

File: feature-runner/steps/device.ts [hunk 20]
```diff
@@ -126,34 +134,37 @@ const getDevice =
...
 const publishDeviceMessage =
-	(bridgeInfo: Settings) =>
+	(nRFCloudSettings: Settings) =>
 	async ({
 		step,
 		log: {
 			step: { progress, error },
 		},
 	}: StepRunnerArgs<World>): Promise<StepRunResult> => {
-		const match =
```
Comment (coderbyheart) [inline feature-runner/steps/device.ts]: Unrelated change.

File: feature-runner/steps/device.ts [hunk 21]
```diff
@@ -126,34 +134,37 @@ const getDevice =
...
+			}),
+		)(
+			/^the device `(?<id>[^`]+)` publishes this message to the topic `(?<topic>[^`]+)`$/,
+			step.title,
+		)
 		if (match === null) return noMatch
 
 		const message = JSON.parse(codeBlockOrThrow(step).code)
 
-		progress(
```
Comment (coderbyheart) [inline feature-runner/steps/device.ts]: Unrelated change.

File: feature-runner/steps/device.ts [hunk 22]
```diff
@@ -162,9 +173,7 @@ const publishDeviceMessage =
 
 			mqttClient.on('connect', () => {
 				progress('connected')
-				const topic = `${bridgeInfo.mqttTopicPrefix}${
```
Comment (coderbyheart) [inline feature-runner/steps/device.ts]: Unrelated change.

File: feature-runner/steps/historicalData.ts [hunk 23]
```diff
@@ -95,10 +95,6 @@ const dateParser = (key: string, value: any) => {
 
 const assertResult = async ({
 	step,
-	log: {
```
Comment (coderbyheart) [inline feature-runner/steps/historicalData.ts]: Unrelated change.

File: feature-runner/steps/storage.ts [hunk 24]
```diff
@@ -34,4 +34,44 @@ const store = async ({
 	context[match.storeName] = result
 }
 
-export const steps = (): StepRunner<World>[] => [store]
+const storeSeries = async ({
```
Comment (coderbyheart) [inline feature-runner/steps/storage.ts]: Remove

File: feature-runner/steps/websocket.ts [hunk 25]
```diff
@@ -18,7 +18,6 @@ import type { World } from '../run-features.js'
 chai.use(chaiSubset)
 
 const wsClients: Record<string, WebSocketClient> = {}
-
```
Comment (coderbyheart) [inline feature-runner/steps/websocket.ts:23]: Unrelated change.

File: features/DeviceShadow.feature.md [hunk 26]
```diff
@@ -326,18 +326,6 @@ Soon I should receive a message on the websocket that matches
       "eest": 7
     }
   },
-  "ts": 1682072423000,
```
Comment (coderbyheart) [inline features/DeviceShadow.feature.md:329]: Unrelated change.

File: historicalData/historicalDataRepository.ts [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+): Record<string, unknown>[] => {
+	const transformedData = []
+
+	// Added ts to mapKeys as default
+	mapKeys.push({ fromKey: 'time', toKey: 'ts' })
+
+	const groupedData = groupBy(data, (d) => d.time)
+	for (const item in groupedData) {
+		const data = groupedData[item]
+		const tranformedRecord = data?.reduce<Record<string, unknown>>(
```
Comment (coderbyheart) [inline historicalData/historicalDataRepository.ts:65]: Please don't mix programming styles, you use both `reduce()` and `for ... in`. Pick one.

File: historicalData/historicalDataRepository.ts [hunk 28]
```diff
@@ -0,0 +1,156 @@
...
+	const [historicalDataDatabaseName, historicalDataTableName] =
+		historicalDataTableInfo.split('|')
+	if (
+		historicalDataDatabaseName === undefined ||
+		historicalDataTableName === undefined
+	) {
+		throw new Error(
+			`Historical data table info must be in format "databaseName|tableName"`,
+		)
+	}
```
Comment (coderbyheart) [inline historicalData/historicalDataRepository.ts]: Move parsing of the table info outside and have the API require table and database name, so it can be reused in other lambdas, and the function here does not need to throw an error.

File: historicalData/queryGenerator.spec.ts [hunk 29]
```diff
@@ -0,0 +1,313 @@
...
+
+import {
+	getAggregates,
+	getBinnedTime,
+	getMeasureNames,
+	getQueryStatement,
+	getStartPeriod,
+} from './queryGenerator.js'
+
+jest.mock('@hello.nrfcloud.com/proto/hello', () => ({
```
Comment (coderbyheart) [inline historicalData/queryGenerator.spec.ts]: Chart types are defined in the backend.

File: historicalData/queryGenerator.spec.ts [hunk 30]
```diff
@@ -0,0 +1,313 @@
...
+}))
+
+describe('queryGenerator', () => {
+	// let request: Omit<Static<typeof HistoricalDataRequest>, 'data'>
+	let request: HistoricalRequest
+	beforeEach(() => {
+		request = {
+			'@context':
+				'https://github.com/hello-nrfcloud/proto/historical-data-request',
+			'@id': 'test-id',
```
Comment (coderbyheart) [inline historicalData/queryGenerator.spec.ts]: use Ulid so it's clear that this is random.

File: lambda/historicalDataRequest.ts [hunk 31]
```diff
@@ -0,0 +1,75 @@
...
+	event: EventBridgeEvent<'request', Request>,
+): Promise<void> => {
+	try {
+		log.info('event', { event })
+
+		const {
+			deviceId,
+			connectionId,
+			message: { request, model },
+		} = event.detail
```
Comment (coderbyheart) [inline lambda/historicalDataRequest.ts:66]: Validate request

**Review by coderbyheart — CHANGES_REQUESTED** [review]

**Comment by coderbyheart** [comment]
I've addressed the unrelated changes regarding to `matchGroups` in https://github.com/hello-nrfcloud/backend/commit/ccdaee3e34b83a32567bb2dd36851729d7e6a01a

File: historicalData/historicalDataRepository.ts (hunk 27 above)
Comment (pudkrong) [reply historicalData/historicalDataRepository.ts:65]: `groupBy` returns `{ [index: string]: T[] }`. So, there is no native `forEach` function to iterate Dictionary

File: lambda/historicalDataRequest.ts (hunk 31 above)
Comment (pudkrong) [reply lambda/historicalDataRequest.ts:66]: This function is triggered by `event bridge`. So, the request has been validated at `onMessage` before sending into event bridge. In addition, there is another validation on `proto` the same approach we use for `convert` message. 

Do we still need validation here?

Comment (coderbyheart) [reply lambda/historicalDataRequest.ts:66]: Ah, yes. In that case we don't need to validate twice.

File: feature-runner/steps/config.ts [hunk 32]
```diff
@@ -65,7 +65,7 @@ export const configStepRunners = ({
 } => ({
 	steps: createConfigStepRunners({ configWriter }),
 	cleanup: async () => {
-		const settings = await configSettings()
+		const settings = await configSettings().catch(() => ({}))
```
Comment (coderbyheart) [inline feature-runner/steps/config.ts]: Don't mix async/await and Promise API

File: feature-runner/lib/codeBlockReplacer.spec.ts [hunk 33]
```diff
@@ -0,0 +1,60 @@
+import { codeBlockReplacer } from './codeBlockReplacer.js'
```
Comment (coderbyheart) [inline feature-runner/lib/codeBlockReplacer.spec.ts]: Replacing data from context is already implemented in BDD Markdown. Use that method.

File: features/HistoricalRequest.feature.md [hunk 34]
```diff
@@ -0,0 +1,303 @@
...
+
+And the device `${fingerprint:deviceId}` publishes this message to the topic
+`m/d/${fingerprint:deviceId}/d2c`
+
+```json
+{
+  "appId": "${appId}",
+  "messageType": "DATA",
+  "data": "${data}",
+  "ts": ${pastTs}
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]:
```suggestion
  "ts": "$number{pastTs}"
```

File: features/HistoricalRequest.feature.md [hunk 35]
```diff
@@ -0,0 +1,303 @@
...
+
+```json
+{
+  "@context": "https://github.com/hello-nrfcloud/proto/historical-data-response",
+  "@id": "420eac59-5ce8-4751-b7d1-217811382095",
+  "attributes": {
+    "avgMA": [
+      {
+        "mA": 3.40141,
+        "ts": `ts`
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]:
```suggestion
        "ts": "$number{ts}"
```

File: features/HistoricalRequest.feature.md [hunk 36]
```diff
@@ -0,0 +1,303 @@
...
+      },
+      {
+        "mA": 3.74543,
+        "ts": `ts - 60000`
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]:
```suggestion
        "ts": "$number{ts - 60000}"
```

File: features/HistoricalRequest.feature.md [hunk 37]
```diff
@@ -0,0 +1,303 @@
...
+      },
+      {
+        "mA": 3.4148300000000003,
+        "ts": `ts - 120000`
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]:
```suggestion
        "ts": "$number{ts - 120000}"
```

File: features/HistoricalRequest.feature.md [hunk 38]
```diff
@@ -0,0 +1,303 @@
...
+
+And the device `${fingerprint:deviceId}` publishes this message to the topic
+`m/d/${fingerprint:deviceId}/d2c`
+
+```json
+{
+  "appId": "${appId}",
+  "messageType": "DATA",
+  "data": "${data}",
+  "ts": ${pastTs}
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]:
```suggestion
  "ts": "$number{pastTs}"
```

File: features/HistoricalRequest.feature.md [hunk 39]
```diff
@@ -0,0 +1,303 @@
...
+
+Given I store `ts - ${deductMsFromTS}` into `pastTs`
+
+And I write Timestream for the device `${fingerprint:deviceId}` with this
+message
+
+```json
+{
+  "@context": "https://github.com/hello-nrfcloud/proto/transformed/PCA20035%2Bsolar/location",
+  "lat": ${lat},
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]:
```suggestion
  "lat": "$number{lat}",
```

File: historicalData/storeRecordsInTimestream.ts [hunk 40]
```diff
@@ -4,27 +4,27 @@ import {
 	WriteRecordsCommand,
 	type _Record,
 } from '@aws-sdk/client-timestream-write'
-import { logger } from '../lambda/util/logger.js'
-
-const log = logger('storeRecordsInTimestream')
+import type { Logger } from '@aws-lambda-powertools/logger'
```
Comment (coderbyheart) [inline historicalData/storeRecordsInTimestream.ts:7]: Unrelated change

**Review by coderbyheart — CHANGES_REQUESTED** [review]

**Comment by coderbyheart** [comment]
The `$number{v}` feature is very new: https://github.com/NordicSemiconductor/bdd-markdown-js/commit/92d4123f2934e4d54496a23d6872a5bdb8567679

File: historicalData/storeRecordsInTimestream.ts (hunk 40 above)
Comment (pudkrong) [reply historicalData/storeRecordsInTimestream.ts:7]: `historicalData/storeRecordsInTimestream.ts` is used in e2e test to store messages into Timestream, https://github.com/hello-nrfcloud/backend/pull/76/files#diff-73d68b7919e0e0232209be295615541da24bff65aceb4bd6eda55499170e54fcR211. Then the logger will print log message and it messes up the feature runner output. Consequently, `console-runner` cannot parse and throw error.

Comment (coderbyheart) [reply historicalData/storeRecordsInTimestream.ts:7]: Ah, I see.

File: features/HistoricalRequest.feature.md [hunk 41]
```diff
@@ -0,0 +1,303 @@
...
+| SOLAR | 3.75718 | 30000          |
+| SOLAR | 3.73368 | 60000          |
+| SOLAR | 3.58041 | 90000          |
+| SOLAR | 3.24925 | 120000         |
+
+## Verify I can query gain historical device data
+
+When I connect to the websocket using fingerprint `${fingerprint}`
+
+And I send websocket request
```
Comment (coderbyheart) [inline features/HistoricalRequest.feature.md]:
```suggestion
And I send this message via the websocket
```

File: feature-runner/steps/websocket.ts [hunk 42]
```diff
@@ -110,11 +110,27 @@ const receive = async ({
...
 }
 
+const wsSend = async ({
+	step,
+	log: {
+		step: { progress },
+	},
+	context,
+}: StepRunnerArgs<World>): Promise<StepRunResult> => {
+	const match = /^I send websocket request$/.exec(step.title)
```
Comment (coderbyheart) [inline feature-runner/steps/websocket.ts]:
```suggestion
	const match = /^I send this message via the websocket$/.exec(step.title)
```

**Review by coderbyheart — CHANGES_REQUESTED** [review]
Please rebase, otherwise it's good to go!
